src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -80,165 +73,77 @@ void Profile::DeleteUnfocusedAppearance()
-            auto copiedUnfocusedAppearance = AppearanceConfig::CopyAppearance(sourceUnfocusedAppearanceImpl, weakRefToProfile);
-
-            // Make sure to add the default appearance as a parent
-            copiedUnfocusedAppearance->InsertParent(copiedDefaultAppearance);
-            unfocused = *copiedUnfocusedAppearance;
-        }
-        profile->_UnfocusedAppearance = unfocused;
+        target.emplace_back(sourceProfile->CopyInterned(visited));
     }
-
-    return profile;
 }
 
-// Method Description:
-// - Creates a copy of the inheritance graph by performing a depth-first traversal recursively.
-//   Profiles are recorded as visited via the "visited" parameter.
-//   Unvisited Profiles are copied into the "cloneGraph" parameter, then marked as visited.
-// Arguments:
-// - sourceGraph - the graph of Profile's we're cloning
-// - cloneGraph - the clone of sourceGraph that is being constructed
-// - visited - a map of which Profiles have been visited, and, if so, a reference to the Profile's clone
-// Return Value:
-// - a clone in both inheritance structure and Profile values of sourceGraph
-winrt::com_ptr<Profile> Profile::CloneInheritanceGraph(winrt::com_ptr<Profile> sourceGraph, winrt::com_ptr<Profile> cloneGraph, std::unordered_map<void*, winrt::com_ptr<Profile>>& visited)
+winrt::com_ptr<Profile>& Profile::CopyInterned(std::unordered_map<const Profile*, winrt::com_ptr<Profile>>& visited) const
 {
-    // If this is an unexplored Profile
-    //   and we have parents...
-    if (visited.find(sourceGraph.get()) == visited.end() && !sourceGraph->_parents.empty())
-    {
-        // iterate through all of our parents to copy them
-        for (const auto& sourceParent : sourceGraph->_parents)
-        {
-            // If we visited this Profile already...
-            auto kv{ visited.find(sourceParent.get()) };
-            if (kv != visited.end())
-            {
-                // add this Profile's clone as a parent
-                InsertParentHelper(cloneGraph, kv->second);
-            }
-            else
-            {
-                // We have not visited this Profile yet,
-                // copy contents of sourceParent to clone
-                winrt::com_ptr<Profile> clone{ CopySettings(sourceParent) };
-
-                // add the new copy to the cloneGraph
-                InsertParentHelper(cloneGraph, clone);
-
-                // copy the sub-graph at "clone"
-                CloneInheritanceGraph(sourceParent, clone, visited);
+    auto& clone = visited[this];
 
-                // mark clone as "visited"
-                // save it to the map in case somebody else references it
-                visited[sourceParent.get()] = clone;
-            }
-        }
+    if (!clone)
+    {
+        clone = CopySettings();
+        CopyInheritanceGraph(visited, _parents, clone->_parents);
+        clone->_FinalizeInheritance();

```

Comment:
Don't we also need to add the clone to `visited` here? That way if we encounter it again, clone will be populated above?

Reply:
or just always `emplace` in `CopyInheritanceGraph`.

Reply:
Ha! This is some evil C++ abuse! 😄
I'm using the `operator[]` a few lines above to get a reference a node in `visited`.
If the node doesn't exist the `operator[]` will default-initialize it (which is an empty `com_ptr`). In this if-condition I'm checking for emptiness and then assign `clone` with a clone. Since `clone` is a reference into the hashmap it writes it into the hashmap directly.